src/data_generators/copula_data_set.py

```diff
@@ -0,0 +1,176 @@
+                ids labels are meaningless while their distribution matters.
+            copula_generator:  An instance of a subclass of
+                statsmodels.distributions.copula.copulas.Copula.  See here for the
+                possible choices:
+                https://github.com/statsmodels/statsmodels/tree/32dc52699f994acbf9dbdb9bd10d7eff04d860f5/statsmodels/distributions
+            universe_size:  A cross-publisher universe size to construct the copula.
+                If not given, then specify the universe size as the maximum single
+                publisher reach times 2.
+            pricing_generator:  A PricingGenerator object that annotates a list of
+                id's with randomly generated price information.  At this moment,
+                assume that all the publishers share the same PricingGenerator.
+            random_generator:  A random generator to draw sample from the copula.
+            name:  If specified, a human-readable name that will be associated to this
+                DataSet.
+        """
+        if universe_size is None:
+            universe_size = 2 * max(
+                [1] + [data.max_reach for data in unlabeled_publisher_data_list]
+            )
+        self.marginal_distributions = []
+        for pub in unlabeled_publisher_data_list:
+            pmf = self.zero_included_pmf(pub, universe_size)
+            self.marginal_distributions.append(
+                stats.rv_discrete(values=(range(len(pmf)), pmf))
+            )
+        self.copula_generator = copula_generator
+        self.distribution = CopulaDistribution(
+            copula=self.copula_generator,
+            marginals=self.marginal_distributions,
+        )
+        self.sample = self.distribution.rvs(
+            nobs=universe_size,
+            random_state=random_generator.integers(0, 1e9),
+        ).astype("int32")
+        imps = self.to_impressions(self.sample)
+        for pub_imps in imps:
+            random_generator.shuffle(pub_imps)
+        super().__init__(
+            publisher_data_list=[
+                PublisherData(
+                    impression_log_data=pricing_generator(pub_imps),
+                    name=original_pub_data.name,
+                )
+                for pub_imps, original_pub_data in zip(
+                    imps, unlabeled_publisher_data_list
+                )
+            ],
+            name=name,
+            universe_size=universe_size,
+        )
+
+    @property
+    def frequency_vectors_sampled_distribution(self) -> Dict:
+        """A dictionary of the distribution of the sampled frequency vectors.
+
+        Returns:
+            A dict d where for any frequency vector (f_1, ..., f_p),
+            d[(f_1, ..., f_p)] is the count of this frequency vector in
+            the DataSet.
+        """
+        return dict(Counter(tuple(obs) for obs in self.sample))
+
+    @staticmethod
+    def zero_included_pmf(
```

Comment:
Suggestion: Shouldn't this be a method of `PublisherData` class instead?

creating a histogram of frequencies starting from zero can be useful other place other then `CopulaDataSet` class.